abci++ spec: remove new_hashes and discuss transaction traceability (#8002)
tendermint/tendermint: spec/abci++/abci++_methods_002_draft.md

    | app_signed_updates      | repeated bytes                                   | Optional changes to the *app_signed* part of vote extensions.                               | 7            |

* **Usage**:
    * The first five parameters of `RequestPrepareProposal` are the same as `RequestProcessProposal`
      and `RequestFinalizeBlock`.
    * The header contains the height, timestamp, and more - it exactly matches the
      Tendermint block header.
    * `RequestPrepareProposal` contains a preliminary set of transactions `txs` that Tendermint considers to be a good block proposal, called _raw block_. The Application can modify this set via `ResponsePrepareProposal.tx_records` (see [TxRecord](#txrecord)).

>         * In this case, the Applications should set `ResponsePrepareProposal.modified_tx` to true.

Reviewer: ```suggestion
        * In this case, the Application should set `ResponsePrepareProposal.modified_tx` to true.
```

        * The Application _can_ reorder, remove or add transactions to the raw block. Let `tx` be a transaction in `txs`:
            * If the Application considers that `tx` should not be proposed in this block, e.g., there are other transactions with higher priority, then it should not include it in `tx_records`. In this case, Tendermint won't remove `tx` from the mempool. The Application should be extra-careful, as abusing this feature may cause transactions to stay forever in the mempool.
            * If the Application considers that a `tx` should not be included in the proposal and removed from the mempool, then the Application should include it in `tx_records` and _mark_ it as "REMOVE". In this case, Tendermint will remove `tx` from the mempool.
            * If the Application wants to add a new transaction, then the Application should include it in `tx_records` and _mark_ it as "ADD". In this case, Tendermint will add it to the mempool.

>         * The Application should be aware that removing and adding transactions may compromise _traceability_. Consider the following example: the Application transforms a client-submitted transaction `t1` into a second transaction `t2`, i.e., the Application asks Tendermint to remove `t1` and add `t2` to the mempool. If a client wants to eventually check what happened to `t1`, it will discover that `t_1` is not in the mempool or in a committed block, getting the wrong idea that `t_1` did not make it into a block. Note that `t_2` _will be_ in a committed block, but unless the Application tracks this information, no component will be aware of it. Thus, if the Application wants traceability, it is its responsability to support it. For instance, the Application could attach to a transformed transaction a list with the hashes of the transactions it derives from. 

Reviewer: über-minor: I would add a line break before "Consider the following example:", making that sentence start a sub-bullet. But up to you!

Author: Done

        * If the Application modifies the set of transactions, the modified transactions MUST NOT exceed the configured maximum size `RequestPrepareProposal.max_tx_bytes`.

>     * If the Application does not modified the premilimnary set of transactions `txs`, then it sets `ResponsePrepareProposal.modified_tx` to false. In this case, Tendermint will ignore the contents of `ResponsePrepareProposal.tx_records`.

Reviewer: ```suggestion
    * If the Application does not modify the preliminary set of transactions `txs`, then it sets `ResponsePrepareProposal.modified_tx` to false. In this case, Tendermint will ignore the contents of `ResponsePrepareProposal.tx_records`.
```

        * remove transactions (invalid) from the proposal and from the mempool - `TxAction = REMOVED`
        * remove transactions from the proposal but not from the mempool (effectively _delaying_ them) - the
          Application removes the transaction from the list

>         * modify transactions (e.g. aggregate them) - `TxAction = ADDED` followed by `TxAction = REMOVED`. As explained above, this compromises client traceability.

Reviewer: ```suggestion
        * modify transactions (e.g. aggregate them) - `TxAction = ADDED` followed by `TxAction = REMOVED`. As explained above, this compromises client traceability, unless it is implemented at the Application level.
```

    |------------|-----------------------|------------------------------------------------------------------|--------------|
    | action     | [TxAction](#txaction) | What should Tendermint do with this transaction?                 | 1            |
    | tx         | bytes                 | Transaction contents                                             | 2            |

>     | new_hashes | repeated bytes        | List of hashes of successor transactions                         | 3            |

Reviewer: Please also remove this from the proto-bufs: proto/tendermint/abci/types.proto
